src/EbtcLeverageZapRouter.sol

```diff
@@ -0,0 +1,263 @@
+        uint256 _debt,
+        bytes32 _upperHint,
+        bytes32 _lowerHint,
+        uint256 _stEthLoanAmount,
+        uint256 _stEthMarginAmount,
+        uint256 _stEthDepositAmount,
+        PositionManagerPermit calldata _positionManagerPermit,
+        bytes calldata _exchangeData
+    ) internal returns (bytes32 cdpId) {
+        // TODO: calculate this for real, need to figure out how to specify leverage ratio
+        // TODO: check max leverage here once we know how leverage will be specified
+        //uint256 flAmount = _debtToCollateral(_debt);
+
+        // We need to deposit slightly less collateral to account for fees / slippage
+        // COLLATERAL_BUFFER is a temporary solution to make the tests pass
+        // TODO: discuss this and see if it's better to pass in some sort of slippage setting
+        //  uint256 totalCollateral = ;
+
+        // TODO: compute CR >= MSCR (minimum safe collateral ratio)
+        // TODO: check fetchPrice gas
+
+        _permitPositionManagerApproval(_positionManagerPermit);
+
+        cdpId = sortedCdps.toCdpId(msg.sender, block.number, sortedCdps.nextCdpNonce());
+
+        OpenCdpForOperation memory cdp;
+
+        cdp.eBTCToMint = _debt;
+        cdp._upperHint = _upperHint;
+        cdp._lowerHint = _lowerHint;
+        cdp.stETHToDeposit = _stEthDepositAmount;
+        cdp.borrower = msg.sender;
+
+        _openCdpOperation({
+            _cdpId: cdpId,
+            _cdp: cdp,
+            _flAmount: _stEthLoanAmount,
+            _stEthBalance: _stEthMarginAmount,
+            _exchangeData: _exchangeData
+        });
+
+        // TODO: emit event
+    }
+
+    function closeCdp(
+        bytes32 _cdpId,
+        PositionManagerPermit calldata _positionManagerPermit,
+        uint256 _stEthAmount,
+        bytes calldata _exchangeData
+    ) external {
+        ICdpManagerData.Cdp memory cdpInfo = cdpManager.Cdps(_cdpId);
+
+        _permitPositionManagerApproval(_positionManagerPermit);
+
+        _closeCdpOperation({
+            _cdpId: _cdpId,
+            _debt: cdpInfo.debt,
+            _stEthAmount: _stEthAmount,
+            _exchangeData: _exchangeData
+        });
+    }
+
+    function _requireMinAdjustment(uint256 _change) internal pure {
+        require(
+            _change >= MIN_CHANGE,
+            "EbtcLeverageZapRouter: Debt or collateral change must be above min"
+        );
+    }
+
+    function _requireZeroOrMinAdjustment(uint256 _change) internal pure {
+        require(
+            _change == 0 || _change >= MIN_CHANGE,
+            "EbtcLeverageZapRouter: Margin increase must be zero or above min"
+        );
+    }
+
+    function adjustCdp(
+        bytes32 _cdpId,
+        AdjustCdpParams calldata params,
+        PositionManagerPermit calldata _positionManagerPermit,
+        bytes calldata _exchangeData
+    ) external {
+        _requireMinAdjustment(params._debtChange);
+        _requireMinAdjustment(params._stEthBalanceChange);
```

Comment:
could user `adjustCdp` by only changing debt? which means `_stEthBalanceChange` might be zero?

Reply:
so the thinking here is that if the user only wants to increase debt without putting any of the proceeds back into the system as additional collateral, then FL is probably not needed and the front-end should route the call to the normal zap router. I am not sure if there's an exception to this?